### ISSUE TITLE
pUML help

### ISSUE THREAD

**Issue by RunjiaChen** [nus-cs2103-AY2324S2/forum#772]
Hello, I am having trouble reaching the intended pUML diagram. 
What I want to draw is something like this:
<img width="775" alt="image" src="https://github.com/nus-cs2103-AY2324S2/forum/assets/142142406/d8d38902-1c9b-4776-b036-4ad4b8b638d8">

It is somewhat similar to this diagram which is on the course website.
<img width="432" alt="image" src="https://github.com/nus-cs2103-AY2324S2/forum/assets/142142406/13f4705c-259d-4ff5-adc2-12689c1c9009">
However, I am only able to get this despite many attempts.
<img width="1191" alt="image" src="https://github.com/nus-cs2103-AY2324S2/forum/assets/142142406/c45bed8a-3ffb-4976-8d24-5b4bc4f89690">
Can anyone provide advice on how to format the UML diagram so that it looks nicer? Thank you!

**Comment by RunjiaChen** [comment]
@damithc Hello Prof, I am still unable to come up with a solution for this problem. May I know how do I proceed? May I know how was the diagram on the website done? Thank you

**Comment by flexibo** [comment]
Hi, I'm a senior from the previous batch.

Can I clarify with you if this activity diagram is only checking if the user input is valid, and then repeating if it is not? 

Based on your diagram, it seems kind of impossible because it looks like you're using 2 [repeats](https://plantuml.com/activity-diagram-beta) at the same time (which is a bit odd since it can either loop or it can't, [you can refer to this for more info](https://plantuml.com/activity-diagram-beta)). Whereas the diagram from the course website is demonstrating condition (which is the 3 arrows going out of the diamond, hence the use of if else here, note that in your drawn diagram there's only 1 outgoing arrow, so the if else is not necessary).

You may want to consider redrawing the diagram, or maybe consider not doing this diagram at all and work on other diagrams if your current diagram is to really demonstrate error handling. Usually, activity diagrams shows higher level, less technical stuff. Hence I think it will be more meaningful to instead show what exactly happens when this user executes this input, what is the main purpose of this AddCommandHelper function? (I'm not very sure what your command is suppose to do but I'm assuming adding a new command, so you can focus on showing this aspect).

Do let me know if you need any further assistance!

**Comment by damithc** [comment]
@RunjiaChen  Combining diamonds may not be possible in PlantUML. You can do something like this:
```
@startuml
start
repeat
repeat
  :p1;
backward: p2; 
repeat while () is ([c1]) not ([not c1])
backward: p3; 
repeat while () is ([c2]) not ([not c2])
stop
@enduml
```

![image](https://github.com/nus-cs2103-AY2324S2/forum/assets/1673303/1903d927-6256-4a57-bdd2-71aacef9f9ec)


> May I know how was the diagram on the website done?

The one you cited is done using PowerPoint.

**Comment by RunjiaChen** [comment]
@flexibo thank you, is there a way i can contact you via email so that I can ask you some questions pertaining to the uml diagrams? Thank you

**Comment by RunjiaChen** [comment]
@flexibo My CommandHelper function actually loads up a separate GUI that prompts users for the fields needed to construct an `add` command. Hence, I split it into 2 diagrams, a diagram that explains the overall flow for the user, and a diagram showing the process for accepting each type of input for the user. 

![Uploading image.png…]()

May I seek your feedback regarding these 2 diagrams? Do you think they are sufficient in explaining my feature? Thank you!

**Comment by flexibo** [comment]
Hi @RunjiaChen SFLR, you can keep commenting in this issue so other people can help too if I'm not available (don't worry I'll keep an eye out on this issue for now, since GitHub will automatically email me if you comment here).

Your image does not seem to load for me btw (it sends me back to this issue). Also are you referring to the `addbystep` command?

**Comment by flexibo** [comment]
@RunjiaChen Ok just took a look at your v1.2 release, I think I understand what you're trying to do now. My suggestion is to show one step at a time instead of generalizing them in your activity diagram. 

So for example: 

## Method 1
```puml
@startuml

start
:AddressBook prompts user for input;
repeat
  :User input name;
repeat while (Valid name?) is (No)
->Yes;
repeat
  :User input phone number;
repeat while (Valid phone number?) is (No)
->Yes;
repeat
  :User input email;
repeat while (Valid email?) is (No)
->Yes;
:AddressBook adds new person;
stop
@enduml
```
![image](https://github.com/nus-cs2103-AY2324S2/forum/assets/40513197/369aeb80-6160-4ad2-81ad-5c0f891f7173)

I understand this method can result in quite a long chain of events (I'm guessing that is why you wanted to condense it). But personally, I still prefer this method as this is a lot clearer and concise. 

## Method 2
I'm not 100% sure if this is legal if someone can confirm for me that will be great, but you can break it down further where one input is an activity (similar to your drawing), 
```puml
@startuml

start
:AddressBook prompts user for input;
repeat
  :User input data;
repeat while (Valid data?) is (No)
->Yes;
:Prompt next input;
stop
@enduml
```
![image](https://github.com/nus-cs2103-AY2324S2/forum/assets/40513197/9079930b-c14e-4978-ab66-9649b83ab7bf)

```puml
@startuml

start
:AddressBook prompts user for the LAST input;
repeat
  :User input data;
repeat while (Valid data?) is (No)
->Yes;
:AddressBook adds new person;
stop
@enduml
```
![image](https://github.com/nus-cs2103-AY2324S2/forum/assets/40513197/b8ae1771-a897-4eff-925a-f404f2bbcdca)

I really don't recommend this method however as this is too general, and also loses some important information, like the required specific parameters.

**Comment by RunjiaChen** [comment]
Hello @flexibo , thank you so much for your feedback! My diagram is closer to method 2 as of now, and it looks something like this: 

![image](https://github.com/nus-cs2103-AY2324S2/forum/assets/142142406/59c51c15-7c13-47e1-874c-f01364ac5ae6)

![image](https://github.com/nus-cs2103-AY2324S2/forum/assets/142142406/ba9f6278-065c-4c1c-a277-525400c2ca84)


I used the rake symbol to link to another diagram, which is the general command for accepting and validating inputs from the user. Do you feel that the diagrams are clear in explaining the flow of the execution? Thank you!

**Comment by flexibo** [comment]
Hi @RunjiaChen, I think your first diagram looks good, so does this mean instead of the app adding in the person, the user will have to copy it with the `cp` command? Your second diagram also looks good, but maybe you can rephrase "Are there more fields needed" (question) to "More fields needed" (condition). Usually you will put the condition within the brackets and question inside the diamond. 


Also idk if you spotted it but "User types the ~~the~~ "cp" command".

**Comment by RunjiaChen** [comment]
Oops, thank you so much for pointing out the error! Regarding putting conditions within the diamond, is it referring to this:
![image](https://github.com/nus-cs2103-AY2324S2/forum/assets/142142406/97c58df9-5176-4e8a-8ff0-6494bfe50f6b)
I can't seem to find it on the 2103 website even though it exists on the puml documentation, so I am worried that i will get penalised for the using non compliant notation. 

Regarding my diagrams, do you think that it would be clearer if I were to add a sequence diagram in explaining the addbystep functionality? Or would it come off as redundant to the reader

**Comment by flexibo** [comment]
@RunjiaChen Yup, I also think it will be safer not to put in the diamond, but I think you should still change the part within square brackets to the condition "More fields needed" instead. This by itself makes it clear and does not need to add the question "Are there more fields needed?" within the diamond anyway. 

I think a sequence diagram will definitely help to explain the `addbystep` functionality, go for it :D!
